day_7/src/main.rs

```diff
@@ -0,0 +1,272 @@
+}
+
+struct Tree<'a> {
+    children: Vec<&'a Tree<'a>>,
+    name: &'a str,
+    weight: u32,
+}
+
+enum ParserState {
+    Name,
+    ExpectBracket,
+    Weight,
+    AnyChildren(usize),
+    ParseChild(usize),
+    ExpectSpace,
+}
+
+fn parse_line(s: &str) -> Node {
+    let mut state = ParserState::Name;
+    let mut count: usize = 0;
+    let mut name_end = 0;
+    let mut weight: u32 = 0;
+    let mut children: Vec<&str> = vec![];
+
+    for chr in s.chars() {
+        match state {
+            ParserState::Name => {
+                if chr == ' ' {
+                    state = ParserState::ExpectBracket;
+                    name_end = count;
+                }
+            }
+            ParserState::ExpectBracket => {
+                if chr != '(' {
+                    panic!("Malformed string! {}", s)
+                }
+                state = ParserState::Weight;
+            }
+            ParserState::Weight => {
+                if chr == ')' {
+                    state = ParserState::AnyChildren(0);
+                } else {
+                    weight = weight * 10
+                        + chr
+                            .to_digit(10)
+                            .unwrap_or_else(|| panic!("Expected a digit, got {}", chr));
+                }
+            }
+            ParserState::AnyChildren(i) => {
+                if i < 3 {
+                    state = ParserState::AnyChildren(i + 1);
+                } else {
+                    state = ParserState::ParseChild(count + 1);
+                }
+            }
+            ParserState::ParseChild(start) => {
+                if chr == ',' {
+                    children.push(&s[start..count]);
+                    state = ParserState::ExpectSpace;
+                }
+            }
+            ParserState::ExpectSpace => {
+                if chr != ' ' {
+                    panic!("Parse failure, expected space, got {}", chr);
+                }
+                state = ParserState::ParseChild(count + 1);
+            }
+        }
+        count += 1;
+    }
+
+    let name = &s[0..name_end];
+    match state {
+        ParserState::AnyChildren(0) => {
+            // No children
+            Node {
+                name,
+                weight,
+                children,
+            }
+        }
+        ParserState::ParseChild(start) => {
+            children.push(&s[start..count]);
+            Node {
+                name,
+                weight,
+                children,
+            }
+        }
+        _ => {
+            panic!("Parse failure");
+        }
+    }
+}
+
+fn input() -> Vec<Node<'static>> {
+    let input = include_str!("../input.txt");
+    input.lines().map(|l| parse_line(l)).collect::<Vec<Node>>()
+}
+
+fn tree_it<'a>(nodes: &[Node<'a>]) -> &'a Tree<'a>
+{
+    let nodes: HashMap<&'a str, &Node<'a>> = nodes.iter().map(|i| (i.name, i)).collect();
+    let mut built: HashMap<&'a str, &'a Tree<'a>> = HashMap::new();
```

Comment:
Exactly one value from this `HashMap` will be directly returned; all the other values are pointing to memory on the heap that is referenced transitively by the one element I return.